scrapy/core/spidermw.py

```diff
@@ -60,11 +60,22 @@ def process_spider_exception(_failure):
             return _failure
 
         def process_spider_output(result):
+            def wrapper(result_iterable):
+                try:
+                    for r in result_iterable:
+                        yield r
+                except Exception as ex:
+                    exception_result = process_spider_exception(Failure(ex))
+                    if exception_result is None or isinstance(exception_result, Failure):
+                        raise
+                    for output in exception_result:
+                        yield output
             for method in self.methods['process_spider_output']:
                 result = method(response=response, result=result, spider=spider)
-                assert _isiterable(result), \
-                    'Middleware %s must returns an iterable object, got %s ' % \
-                    (fname(method), type(result))
+                if _isiterable(result):
```

Comment:
@kmike [This test](https://github.com/scrapy/scrapy/pull/2061/files#diff-442dbe445eb9f6625bfb8186556983f2R269) ensures only non already executed `process_spider_output` methods are called when `process_spider_exception` returns an iterable, breaking the exception handling chain :-)